File: weird_captcha_gym/shared_runtime/app/mechanics/statute_yard.js
(() => {
  "use strict";

  const SUBJECTS = new Set(["STATUE", "LANTERN", "GATE", "CISTERN"]);
  const PROPERTIES = new Set(["YOU", "EXIT", "DEADLY", "STOP"]);
  const DIRECTIONS = {
    UP: [0, -1], RIGHT: [1, 0], DOWN: [0, 1], LEFT: [-1, 0],
  };
  const KEYS = {
    arrowup: "UP", w: "UP", arrowright: "RIGHT", d: "RIGHT",
    arrowdown: "DOWN", s: "DOWN", arrowleft: "LEFT", a: "LEFT",
  };
  let model = null;
  let activeCleanup = null;

  function pointKey(x, y) { return `${Number(x)},${Number(y)}`; }
  function clonePositions(items) { return items.map((item) => ({id: String(item.id), x: Number(item.x), y: Number(item.y)})); }
  function positionMap(items) { return new Map(items.map((item, index) => [pointKey(item.x, item.y), index])); }

  function activeRules() {
    const at = new Map(model.words.map((word, index) => [pointKey(word.x, word.y), model.yard.words[index].text]));
    const rules = new Set();
    model.words.forEach((word, index) => {
      if (model.yard.words[index].text !== "IS") return;
      const x = Number(word.x);
      const y = Number(word.y);
      [
        [at.get(pointKey(x - 1, y)), at.get(pointKey(x + 1, y))],
        [at.get(pointKey(x, y - 1)), at.get(pointKey(x, y + 1))],
      ].forEach(([subject, property]) => {
        if (SUBJECTS.has(subject) && PROPERTIES.has(property)) rules.add(`${subject} IS ${property}`);
      });
    });
    return [...rules].sort();
  }

  function kindsWith(rules, property) {
    const suffix = ` IS ${property}`;
    return new Set(rules.filter((rule) => rule.endsWith(suffix)).map((rule) => rule.slice(0, -suffix.length)));
  }

  function snapshot() {
    const rules = activeRules();
    return {
      entities: clonePositions(model.entities),
      words: clonePositions(model.words),
      active_rules: rules,
      controlled_kinds: [...kindsWith(rules, "YOU")].sort(),
      broken_opening_rules: [...model.broken].sort(),
    };
  }

  function predicateState(rules) {
    const controlled = kindsWith(rules, "YOU");
    const exits = kindsWith(rules, "EXIT");
    const deadly = kindsWith(rules, "DEADLY");
    const controlledCells = new Set(
      model.entities.filter((entity, index) => controlled.has(model.yard.entities[index].kind)).map((entity) => pointKey(entity.x, entity.y)),
    );
    const won = controlledCells.size > 0 && model.entities.some(
      (entity, index) => exits.has(model.yard.entities[index].kind) && controlledCells.has(pointKey(entity.x, entity.y)),
    );
    const dead = model.entities.some(
      (entity, index) => deadly.has(model.yard.entities[index].kind) && controlledCells.has(pointKey(entity.x, entity.y)),
    );
    return {won, dead};
  }

  function record(event) {
    model.actions.push({seq: model.actions.length + 1, ...event});
  }

  function clearTransientVerdict() {
    const root = document.querySelector(".statute-yard");
    if (!root?.classList.contains("is-failed")) return;
    root.classList.remove("is-failed");
    const panel = document.querySelector(".statute-verdict");
    if (panel) panel.replaceChildren();
  }

  function move(direction, inputSource) {
    if (!model || model.terminal || model.submitting || !(direction in DIRECTIONS)) return;
    clearTransientVerdict();
    const [dx, dy] = DIRECTIONS[direction];
    const beforeRules = activeRules();
    const controlled = kindsWith(beforeRules, "YOU");
    const stopped = kindsWith(beforeRules, "STOP");
    const controlledIndices = model.yard.entities
      .map((entity, index) => controlled.has(entity.kind) ? index : -1)
      .filter((index) => index >= 0);
    let pushed = false;
    let moved = false;

    controlledIndices.forEach((entityIndex) => {
      const entity = model.entities[entityIndex];
      const target = {x: entity.x + dx, y: entity.y + dy};
      if (model.walls.has(pointKey(target.x, target.y))) return;
      let wordByCell = positionMap(model.words);
      const firstWord = wordByCell.get(pointKey(target.x, target.y));
      if (firstWord !== undefined) {
        const chain = [];
        let cursor = {x: target.x, y: target.y};
        while (wordByCell.has(pointKey(cursor.x, cursor.y))) {
          chain.push(wordByCell.get(pointKey(cursor.x, cursor.y)));
          cursor = {x: cursor.x + dx, y: cursor.y + dy};
        }
        const objectStops = model.entities.some((other, index) =>
          other.x === cursor.x && other.y === cursor.y && stopped.has(model.yard.entities[index].kind));
        if (model.walls.has(pointKey(cursor.x, cursor.y)) || objectStops) return;
        [...chain].reverse().forEach((wordIndex) => {
          model.words[wordIndex].x += dx;
          model.words[wordIndex].y += dy;
        });
        pushed = true;
        wordByCell = positionMap(model.words);
      }
      const targetStopped = model.entities.some((other, index) =>
        other.x === target.x && other.y === target.y && stopped.has(model.yard.entities[index].kind));
      if (targetStopped) return;
      entity.x = target.x;
      entity.y = target.y;
      moved = true;
    });

    const afterRules = activeRules();
    model.openingRules.forEach((rule) => { if (!afterRules.includes(rule)) model.broken.add(rule); });
    const predicates = predicateState(afterRules);
    model.won = !predicates.dead && predicates.won;
    model.dead = predicates.dead;
    model.terminal = model.won || model.dead;
    let outcome = "blocked";
    if (model.dead) outcome = "deadly_contact";
    else if (model.won) outcome = "exit_reached";
    else if (pushed) outcome = "law_shift";
    else if (moved) outcome = "move";
    else if (controlledIndices.length === 0) outcome = "no_subject_bound";
    record({type: "move", direction, input_source: inputSource, outcome});
    renderState(outcome);
  }

  function resetYard() {
    if (!model || model.terminal || model.submitting) return;
    clearTransientVerdict();
    model.entities = clonePositions(model.yard.entities);
    model.words = clonePositions(model.yard.words);
    model.broken = new Set();
    model.resetCount += 1;
    model.won = false;
    model.dead = false;
    const inputSource = model.interaction === "full" ? "keyboard" : "direction_buttons";
    record({type: "reset", input_source: inputSource, outcome: "reset"});
    renderState("reset");
  }

  function statusCopy(outcome, rules) {
    if (model.dead) return ["DEADLY CONTACT", "error"];
    if (model.won) return ["EXIT PREDICATE TRUE · SEAL THE VERDICT", "passed"];
    if (outcome === "law_shift") return [`LAW SHIFT · ${rules.length} ACTIVE`, "idle"];
    if (outcome === "blocked") return ["STONEWORK BLOCKED", "error"];
    if (outcome === "no_subject_bound") return ["NO OBJECT IS YOU", "error"];
    if (outcome === "reset") return ["YARD RESTORED · TRANSCRIPT RETAINED", "idle"];
    return ["MOVEMENT ENTERED", "idle"];
  }

  function tokenMarkup(kind, count, index) {
    const labels = {STATUE: "S", LANTERN: "L", GATE: "G", CISTERN: "C"};
    const extras = count > 1 ? `<i>${count}</i>` : "";
    const shiftX = index * 5;
    const shiftY = index * -4;
    return `<span class="yard-object object-${kind.toLowerCase()}" style="--object-shift-x:${shiftX}px;--object-shift-y:${shiftY}px"><b>${labels[kind]}</b>${extras}</span>`;
  }

  function renderBoard() {
    const rules = activeRules();
    const controlled = kindsWith(rules, "YOU");
    const exits = kindsWith(rules, "EXIT");
    const deadly = kindsWith(rules, "DEADLY");
    const wordAt = new Map(model.words.map((word, index) => [pointKey(word.x, word.y), model.yard.words[index]]));
    const cells = [];
    for (let y = 0; y < Number(model.yard.height); y += 1) {
      for (let x = 0; x < Number(model.yard.width); x += 1) {
        const key = pointKey(x, y);
        const wall = model.walls.has(key);
        const crack = model.cracks.has(key);
        const word = wordAt.get(key);
        const objects = model.entities
          .map((entity, index) => ({entity, kind: model.yard.entities[index].kind}))
          .filter((entry) => entry.entity.x === x && entry.entity.y === y);
        const grouped = new Map();
        objects.forEach((entry) => grouped.set(entry.kind, (grouped.get(entry.kind) || 0) + 1));
        const objectMarkup = [...grouped].map(([kind, count], index) => tokenMarkup(kind, count, index)).join("");
        const flags = objects.map((entry) => entry.kind);
        cells.push(`<div class="yard-cell${wall ? " is-wall" : ""}${crack ? " is-cracked" : ""}${word ? " has-word" : ""}${flags.some((kind) => controlled.has(kind)) ? " has-you" : ""}${flags.some((kind) => exits.has(kind)) ? " has-exit" : ""}${flags.some((kind) => deadly.has(kind)) ? " has-deadly" : ""}" data-cell="${key}">
          ${word ? `<span class="law-stone word-${word.text.toLowerCase()}"><i></i><b>${model.helpers.text(word.text)}</b></span>` : ""}
          ${objectMarkup}
        </div>`);
      }
    }
    const board = document.querySelector(".statute-board");
    if (board) board.innerHTML = cells.join("");
    document.querySelectorAll(".rule-ledger-list").forEach((ledger) => {
      ledger.innerHTML = rules.map((rule) => {
        const brokenOpening = model.openingRules.has(rule) ? " opening" : " enacted";
        return `<li class="${brokenOpening.trim()}"><i></i><span>${model.helpers.text(rule)}</span><b>ACTIVE</b></li>`;
      }).join("") || "<li class=\"empty\">NO LAW RESOLVES</li>";
    });
    const broken = document.querySelector(".broken-law-count b");
    if (broken) broken.textContent = String(model.broken.size);
    const moves = document.querySelector(".yard-move-count b");
    if (moves) moves.textContent = String(model.actions.filter((action) => action.type === "move").length).padStart(2, "0");
    const identity = document.querySelector(".current-you b");
    if (identity) identity.textContent = [...controlled].sort().join(" + ") || "NONE";
    document.querySelectorAll(".yard-direction, .yard-reset, .yard-seal").forEach((button) => {
      button.disabled = model.submitting;
    });
    const root = document.querySelector(".statute-yard");
    root?.classList.toggle("is-won", model.won);
    root?.classList.toggle("is-dead", model.dead);
  }

  function renderState(outcome = "initial") {
    renderBoard();
    const rules = activeRules();
    const [message, status] = statusCopy(outcome, rules);
    model.helpers.setReadout(message, status);
  }

  function verdict(kind, title, detail) {
    const root = document.querySelector(".statute-yard");
    const panel = document.querySelector(".statute-verdict");
    if (!root || !panel) return;
    root.classList.toggle("is-passed", kind === "pass");
    root.classList.toggle("is-failed", kind === "fail");
    panel.innerHTML = `<b>${title}</b><span>${detail}</span>`;
  }

  async function sealVerdict() {
    if (!model || model.submitting) return;
    model.submitting = true;
    renderBoard();
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      completed: model.won === true,
      actions: model.actions,
      reset_count: model.resetCount,
      final_state: snapshot(),
    };
    try {
      const response = await fetch("/result", {
        method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify(payload),
      });
      const outcome = await response.json();
      if (outcome.passed === true) {
        verdict("pass", "RATIFIED", "PASS · EXIT LAW SATISFIED");
        model.helpers.setReadout("PASS · STATUTE RATIFIED", "passed");
      } else if (outcome.state) {
        const helpers = model.helpers;
        await render(outcome.state, helpers);
        verdict("fail", "REPEALED", "FAIL · FRESH YARD ISSUED");
        model.helpers.setReadout("FAIL · FRESH YARD ISSUED", "error");
      } else {
        model.submitting = false;
        verdict("fail", "REJECTED", "FAIL · VERDICT DID NOT VERIFY");
        model.helpers.setReadout("FAIL · VERDICT DID NOT VERIFY", "error");
      }
    } catch (_error) {
      model.submitting = false;
      model.helpers.setReadout("CLERK LINK LOST", "error");
      renderBoard();
    }
  }

  function installDeveloperReveal() {
    const form = document.getElementById("cheat-form");
    const input = document.getElementById("cheat-password");
    const output = document.getElementById("cheat-output");
    if (!form || !input || !output) return;
    form.addEventListener("submit", async (event) => {
      event.preventDefault();
      const response = await fetch("/cheat", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify({password: input.value})});
      if (!response.ok) { output.textContent = response.status === 404 ? "Disabled." : "Denied."; return; }
      const data = await response.json();
      output.textContent = `Route: ${(data.route || []).join(" ")} · minimum ${(data.minimum_solution_steps || 0)}`;
    });
  }

  async function render(state, helpers) {
    if (activeCleanup) activeCleanup();
    document.body.dataset.mechanic = "statute-yard";
    document.body.dataset.cheatMode = helpers.isCheatMode() ? "true" : "false";
    const yard = state.yard;
    const interaction = state.control_condition?.interaction || "full";
    model = {
      state, helpers, yard, interaction,
      entities: clonePositions(yard.entities), words: clonePositions(yard.words),
      walls: new Set((yard.walls || []).map((point) => pointKey(point[0], point[1]))),
      cracks: new Set((yard.cracks || []).map((point) => pointKey(point[0], point[1]))),
      openingRules: new Set(state.opening_rules || []), broken: new Set(),
      actions: [], resetCount: 0, won: false, dead: false, terminal: false, submitting: false,
    };
    helpers.app.innerHTML = `
      <section class="statute-yard palette-${helpers.text(yard.palette || "verdigris")}" data-interaction="${helpers.text(interaction)}" tabindex="0">
        <div class="statute-verdict" aria-live="assertive"></div>
        <header class="statute-head">
          <div><span>OFFICE OF MUTABLE ORDINANCES · YARD ${helpers.text(state.challenge_id)}</span><h1>${helpers.text(state.prompt)}</h1></div>
          <div class="yard-seal-mark"><i>SY</i><b>STATUTE<br>YARD</b></div>
        </header>
        <main class="statute-workbench">
          <section class="yard-map-panel">
            <div class="yard-map-label"><span>REGISTERED GROUNDS</span><b>${Number(yard.width)} × ${Number(yard.height)} · ${helpers.text(String(yard.orientation || "landscape").toUpperCase())}</b></div>
            <div class="statute-board" style="--yard-columns:${Number(yard.width)};--yard-rows:${Number(yard.height)};--yard-object-size:${Number(yard.height) > 10 ? 26 : 42}px"></div>
            <div class="yard-map-footer">
              ${interaction === "simplified" ? `<div class="yard-directions" aria-label="Direction controls">
                <button class="yard-direction" type="button" data-direction="UP" aria-label="Move up">↑</button>
                <button class="yard-direction" type="button" data-direction="LEFT" aria-label="Move left">←</button>
                <button class="yard-direction" type="button" data-direction="DOWN" aria-label="Move down">↓</button>
                <button class="yard-direction" type="button" data-direction="RIGHT" aria-label="Move right">→</button>
                <button class="yard-reset" type="button">RESET</button>
              </div>` : `<span class="key-legend">ARROWS / WASD TO MOVE · R TO RESET</span>`}
              <b>WORD-STONES PUSH · LAWS RECOMPUTE AFTER EVERY MOVE</b>
            </div>
          </section>
          <aside class="statute-ledger">
            <div class="ledger-title"><span>ACTIVE LAW LEDGER</span><i>LIVE</i></div>
            <ol class="rule-ledger-list"></ol>
            <div class="law-syntax"><span>VALID FORM</span><b>NOUN</b><i>IS</i><b>PROPERTY</b><small>horizontal or vertical</small></div>
            <div class="yard-tallies">
              <span class="current-you">YOU <b>—</b></span>
              <span class="broken-law-count">OPENING LAWS BROKEN <b>0</b></span>
              <span class="yard-move-count">MOVES <b>00</b></span>
            </div>
          </aside>
        </main>
        <footer class="statute-foot"><div class="readout" data-status="idle">YARD ENTERED · OPENING LAWS REGISTERED</div><button class="yard-seal" type="button">${helpers.text(state.submit_label || "SEAL VERDICT")}</button></footer>
        ${helpers.cheatPanelTemplate()}
      </section>`;

    const keydown = (event) => {
      if (!model || model.interaction !== "full" || event.repeat) return;
      const key = String(event.key || "").toLowerCase();
      if (key === "r") { event.preventDefault(); resetYard(); return; }
      const direction = KEYS[key];
      if (!direction) return;
      event.preventDefault();
      move(direction, "keyboard");
    };
    window.addEventListener("keydown", keydown);
    document.querySelectorAll(".yard-direction").forEach((button) => button.addEventListener("click", () => move(button.dataset.direction, "direction_buttons")));
    document.querySelector(".yard-reset")?.addEventListener("click", resetYard);
    document.querySelector(".yard-seal")?.addEventListener("click", sealVerdict);
    activeCleanup = () => window.removeEventListener("keydown", keydown);
    installDeveloperReveal();
    renderState("initial");
    document.querySelector(".statute-yard")?.focus();
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.statute_yard = {rootSelector: ".statute-yard", render};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/statute_yard.py
from __future__ import annotations

from typing import Any


MECHANIC_ID = "statute_yard"
PROFILE_NAMES = {
    1: "yard_l1",
    2: "yard_l2",
    3: "yard_original_l4",
    4: "yard_l4",
    5: "yard_l5",
}
DIRECTIONS = {
    "UP": (0, -1),
    "RIGHT": (1, 0),
    "DOWN": (0, 1),
    "LEFT": (-1, 0),
}
SUBJECTS = {"STATUE", "LANTERN", "GATE", "CISTERN"}
PROPERTIES = {"YOU", "EXIT", "DEADLY", "STOP"}


def _fail(feedback: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "score": 0, "feedback": feedback}


def _point(item: Any) -> tuple[int, int] | None:
    if not isinstance(item, dict):
        return None
    x, y = item.get("x"), item.get("y")
    if isinstance(x, bool) or isinstance(y, bool) or not isinstance(x, int) or not isinstance(y, int):
        return None
    return x, y


def _validate_yard(yard: Any) -> str | None:
    if not isinstance(yard, dict):
        return "yard is missing"
    width, height = yard.get("width"), yard.get("height")
    if isinstance(width, bool) or isinstance(height, bool) or not isinstance(width, int) or not isinstance(height, int):
        return "yard dimensions are malformed"
    if not 7 <= width <= 14 or not 7 <= height <= 14:
        return "yard dimensions are outside limits"
    walls_raw = yard.get("walls")
    entities = yard.get("entities")
    words = yard.get("words")
    if not isinstance(walls_raw, list) or not isinstance(entities, list) or not isinstance(words, list):
        return "yard collections are malformed"
    walls: set[tuple[int, int]] = set()
    for value in walls_raw:
        if not isinstance(value, list) or len(value) != 2 or not all(isinstance(v, int) and not isinstance(v, bool) for v in value):
            return "wall coordinate is malformed"
        walls.add((value[0], value[1]))
    if len(walls) != len(walls_raw):
        return "wall coordinates are duplicated"
    expected_border = {
        (x, y)
        for y in range(height)
        for x in range(width)
        if x in {0, width - 1} or y in {0, height - 1}
    }
    if not expected_border <= walls:
        return "yard border is open"
    entity_ids: set[str] = set()
    for entity in entities:
        identifier = str(entity.get("id") or "") if isinstance(entity, dict) else ""
        kind = str(entity.get("kind") or "") if isinstance(entity, dict) else ""
        position = _point(entity)
        if not identifier or identifier in entity_ids or kind not in SUBJECTS or position is None:
            return "yard entity is malformed"
        if position in walls:
            return "yard entity begins in a wall"
        entity_ids.add(identifier)
    word_ids: set[str] = set()
    word_positions: set[tuple[int, int]] = set()
    for word in words:
        identifier = str(word.get("id") or "") if isinstance(word, dict) else ""
        text = str(word.get("text") or "") if isinstance(word, dict) else ""
        position = _point(word)
        if (
            not identifier
            or identifier in word_ids
            or text not in SUBJECTS | PROPERTIES | {"IS"}
            or position is None
            or position in walls
            or position in word_positions
        ):
            return "yard word-stone is malformed"
        word_ids.add(identifier)
        word_positions.add(position)
    if not 6 <= len(words) <= 18 or not 2 <= len(entities) <= 12:
        return "yard object count is outside limits"
    return None


def _active_rules(words: list[dict[str, Any]], positions: list[tuple[int, int]]) -> tuple[str, ...]:
    at = {position: words[index]["text"] for index, position in enumerate(positions)}
    rules: set[str] = set()
    for index, word in enumerate(words):
        if word["text"] != "IS":
            continue
        x, y = positions[index]
        for before, after in (((x - 1, y), (x + 1, y)), ((x, y - 1), (x, y + 1))):
            subject, prop = at.get(before), at.get(after)
            if subject in SUBJECTS and prop in PROPERTIES:
                rules.add(f"{subject} IS {prop}")
    return tuple(sorted(rules))


def _controlled(rules: tuple[str, ...]) -> set[str]:
    return {rule.removesuffix(" IS YOU") for rule in rules if rule.endswith(" IS YOU")}


def _properties(rules: tuple[str, ...], prop: str) -> set[str]:
    suffix = f" IS {prop}"
    return {rule.removesuffix(suffix) for rule in rules if rule.endswith(suffix)}


def _snapshot(
    yard: dict[str, Any],
    entity_positions: list[tuple[int, int]],
    word_positions: list[tuple[int, int]],
    broken: set[str],
) -> dict[str, Any]:
    rules = _active_rules(yard["words"], word_positions)
    return {
        "entities": [
            {"id": entity["id"], "x": position[0], "y": position[1]}
            for entity, position in zip(yard["entities"], entity_positions)
        ],
        "words": [
            {"id": word["id"], "x": position[0], "y": position[1]}
            for word, position in zip(yard["words"], word_positions)
        ],
        "active_rules": list(rules),
        "controlled_kinds": sorted(_controlled(rules)),
        "broken_opening_rules": sorted(broken),
    }


def _won(yard: dict[str, Any], entity_positions: list[tuple[int, int]], rules: tuple[str, ...]) -> bool:
    controlled = _controlled(rules)
    exits = _properties(rules, "EXIT")
    controlled_positions = {
        position
        for entity, position in zip(yard["entities"], entity_positions)
        if entity["kind"] in controlled
    }
    return bool(controlled_positions) and any(
        entity["kind"] in exits and position in controlled_positions
        for entity, position in zip(yard["entities"], entity_positions)
    )


def _deadly(yard: dict[str, Any], entity_positions: list[tuple[int, int]], rules: tuple[str, ...]) -> bool:
    controlled = _controlled(rules)
    hazards = _properties(rules, "DEADLY")
    controlled_positions = {
        position
        for entity, position in zip(yard["entities"], entity_positions)
        if entity["kind"] in controlled
    }
    return any(
        entity["kind"] in hazards and position in controlled_positions
        for entity, position in zip(yard["entities"], entity_positions)
    )


def _move(
    yard: dict[str, Any],
    entity_positions: list[tuple[int, int]],
    word_positions: list[tuple[int, int]],
    direction: str,
) -> tuple[list[tuple[int, int]], list[tuple[int, int]], tuple[str, ...], str]:
    dx, dy = DIRECTIONS[direction]
    walls = {tuple(point) for point in yard["walls"]}
    rules_before = _active_rules(yard["words"], word_positions)
    controlled = _controlled(rules_before)
    stopped = _properties(rules_before, "STOP")
    next_entities = list(entity_positions)
    next_words = list(word_positions)
    pushed = False
    moved = False
    controlled_indices = [
        index for index, entity in enumerate(yard["entities"]) if entity["kind"] in controlled
    ]
    for entity_index in controlled_indices:
        x, y = next_entities[entity_index]
        target = (x + dx, y + dy)
        if target in walls:
            continue
        word_index = next((index for index, position in enumerate(next_words) if position == target), None)
        if word_index is not None:
            chain: list[int] = []
            cursor = target
            while True:
                found = next((index for index, position in enumerate(next_words) if position == cursor), None)
                if found is None:
                    break
                chain.append(found)
                cursor = (cursor[0] + dx, cursor[1] + dy)
            if cursor in walls or any(
                position == cursor and yard["entities"][index]["kind"] in stopped
                for index, position in enumerate(next_entities)
            ):
                continue
            for index in reversed(chain):
                wx, wy = next_words[index]
                next_words[index] = (wx + dx, wy + dy)
            pushed = True
        if any(
            position == target and yard["entities"][index]["kind"] in stopped
            for index, position in enumerate(next_entities)
        ):
            continue
        next_entities[entity_index] = target
        moved = True
    rules_after = _active_rules(yard["words"], next_words)
    if _deadly(yard, next_entities, rules_after):
        outcome = "deadly_contact"
    elif _won(yard, next_entities, rules_after):
        outcome = "exit_reached"
    elif pushed:
        outcome = "law_shift"
    elif moved:
        outcome = "move"
    elif not controlled_indices:
        outcome = "no_subject_bound"
    else:
        outcome = "blocked"
    return next_entities, next_words, rules_after, outcome


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    if str(payload.get("mechanic_id") or "") != MECHANIC_ID:
        return _fail("mechanic mismatch")
    if str(ground_truth.get("mechanic_id") or "") != MECHANIC_ID or str(public_state.get("mechanic_id") or "") != MECHANIC_ID:
        return _fail("statute-yard contract mechanic mismatch")
    challenge_id = str(ground_truth.get("challenge_id") or "")
    if not challenge_id or str(payload.get("challenge_id") or "") != challenge_id:
        return _fail("stale challenge")
    if str(public_state.get("challenge_id") or "") != challenge_id:
        return _fail("public challenge mismatch")
    task_id = str(ground_truth.get("task_id") or "")
    if not task_id or str(payload.get("task_id") or "") != task_id or str(public_state.get("task_id") or "") != task_id:
        return _fail("task identity mismatch")
    yard = ground_truth.get("yard")
    if yard != public_state.get("yard"):
        return _fail("public and hidden yard geometry disagree")
    yard_error = _validate_yard(yard)
    if yard_error:
        return _fail(f"invalid yard contract: {yard_error}")
    truth_condition = ground_truth.get("control_condition")
    if truth_condition != public_state.get("control_condition"):
        return _fail("public interaction condition differs from yard contract")
    interaction = str((truth_condition or {}).get("interaction") or "full")
    expected_source = {"simplified": "direction_buttons", "full": "keyboard"}.get(interaction)
    if expected_source is None:
        return _fail("invalid statute-yard interaction condition")
    parameters = dict((truth_condition or {}).get("difficulty_parameters") or {})
    level = int((truth_condition or {}).get("difficulty") or ground_truth.get("profile") or 3)
    if level not in PROFILE_NAMES or (parameters and parameters.get("profile") != PROFILE_NAMES[level]):
        return _fail("difficulty profile differs from yard contract")

    actions = payload.get("actions")
    if not isinstance(actions, list) or not 1 <= len(actions) <= 500:
        return _fail("movement transcript is missing or outside limits")
    opening_rules = set(ground_truth.get("opening_rules") or [])
    if not opening_rules or opening_rules != set(public_state.get("opening_rules") or []):
        return _fail("opening rule ledger is malformed")
    initial_entities = [_point(entity) for entity in yard["entities"]]
    initial_words = [_point(word) for word in yard["words"]]
    if any(point is None for point in initial_entities + initial_words):
        return _fail("initial yard coordinates are malformed")
    entity_positions = [point for point in initial_entities if point is not None]
    word_positions = [point for point in initial_words if point is not None]
    broken: set[str] = set()
    resets = 0
    terminal = False
    won = False
    deadly = False

    for index, event in enumerate(actions, start=1):
        if not isinstance(event, dict) or event.get("seq") != index:
            return _fail(f"action {index} has an invalid sequence")
        if terminal:
            return _fail("transcript continues after a terminal yard state")
        action_type = str(event.get("type") or "")
        if action_type == "reset":
            if event.get("input_source") != expected_source:
                return _fail(f"action {index} uses the wrong interaction input")
            entity_positions = list(initial_entities)  # type: ignore[arg-type]
            word_positions = list(initial_words)  # type: ignore[arg-type]
            broken = set()
            resets += 1
            expected_outcome = "reset"
        elif action_type == "move":
            if event.get("input_source") != expected_source:
                return _fail(f"action {index} uses the wrong interaction input")
            direction = str(event.get("direction") or "").upper()
            if direction not in DIRECTIONS:
                return _fail(f"action {index} has an invalid direction")
            entity_positions, word_positions, rules, expected_outcome = _move(
                yard, entity_positions, word_positions, direction
            )
            broken |= opening_rules - set(rules)
            deadly = expected_outcome == "deadly_contact"
            won = expected_outcome == "exit_reached"
            terminal = deadly or won
        else:
            return _fail(f"action {index} has an invalid type")
        if event.get("outcome") != expected_outcome:
            return _fail(
                f"action {index} outcome differs from replay: expected {expected_outcome!r}, got {event.get('outcome')!r}"
            )

    final_snapshot = _snapshot(yard, entity_positions, word_positions, broken)
    if payload.get("final_state") != final_snapshot:
        return _fail("submitted final yard does not match independent replay")
    if payload.get("reset_count") != resets:
        return _fail("submitted reset count does not match replay")
    required_breaks = int(ground_truth.get("required_opening_law_breaks") or 1)
    completed = payload.get("completed") is True
    passed = completed and won and not deadly and len(broken) >= required_breaks
    return {
        "graded": True,
        "passed": passed,
        "score": 100 if passed else 0,
        "feedback": (
            f"active laws {len(final_snapshot['active_rules'])}; opening laws broken {len(broken)}/{required_breaks}; "
            f"moves {sum(1 for action in actions if action.get('type') == 'move')}; resets {resets}; "
            f"exit predicate {'true' if won else 'false'}"
        ),
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    return {
        "route": list(ground_truth.get("solution") or []),
        "minimum_solution_steps": ground_truth.get("minimum_solution_steps"),
        "instruction": "Issue each direction in order, then seal the verdict.",
        "answers": [],
    }
